File: examples/Actor.Next/01-Cart/Cart.Next.Example01.Tests/CartActorTests.cs
// ------------------------------------------------------------------------
// Copyright 2026 The Dapr Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

using Dapr.Actors.Next.Abstractions;
using Dapr.Actors.Next.Abstractions.Options;
using Dapr.Actors.Next.Testing;
using Microsoft.Extensions.DependencyInjection;

namespace Dapr.Actors.Next.Examples.Cart.Tests;

public sealed class CartActorTests
{
    [Fact]
    public async Task Adding_an_item_updates_the_summary()
    {
        await using var runtime = CreateRuntime();
        var cart = runtime.CreateActor<ICartActor>(ActorId.Create("cart-1"), "Cart");

        var add = cart.AddItem(new CartItem("sku-1", 2));
        await runtime.RunToIdle();
        await add;

        var read = cart.GetSummary();
        await runtime.RunToIdle();
        var summary = await read;

        Assert.Equal(2, summary.ItemCount);
        Assert.Equal(25.00m, summary.Total);
        Assert.False(summary.Abandoned);
    }

    [Fact]
    public async Task Advancing_virtual_time_fires_the_abandon_timer()
    {
        await using var runtime = CreateRuntime();
        var cart = runtime.CreateActor<ICartActor>(ActorId.Create("idle-cart"), "Cart");

        var add = cart.AddItem(new CartItem("sku-1", 1));
        await runtime.RunToIdle();
        await add;

        runtime.Time.Advance(TimeSpan.FromMinutes(20));
        await runtime.RunToIdle();

        var summaryTask = cart.GetSummary();
        await runtime.RunToIdle();
        Assert.True((await summaryTask).Abandoned);
    }

    private static ActorTestRuntime CreateRuntime()
    {
        _ = typeof(CartActor);
        return new ActorTestRuntime(services =>
        {
            services.AddSingleton<IPricingClient, FakePricingClient>();
            services.AddDaprActors(_ => { });
        });
    }

    private sealed class FakePricingClient : IPricingClient
    {
        public ValueTask<decimal> GetPriceAsync(string sku, CancellationToken cancellationToken = default) =>
            ValueTask.FromResult(sku == "sku-1" ? 12.50m : 1.00m);
    }
}

File: examples/Actor.Next.FSharp/01-Cart/Cart.Next.FSharp.Example01.Tests/CartActorTests.fs
namespace Cart.Next.FSharp.Example01.Tests

open System
open System.Threading
open System.Threading.Tasks
open Dapr.Actors.Next.Abstractions
open Dapr.Actors.Next.Abstractions.Options
open Dapr.Actors.Next.Testing
open Microsoft.Extensions.DependencyInjection
open Cart.Next.FSharp.Example01
open Xunit

type CartActorTests() =

    static member private CreateRuntime() =
        // Force-load the Glue assembly and invoke its generated registration module
        // (the module initializer does not fire for LoadFrom in this runtime version)
        let gluePath = System.IO.Path.Combine(AppContext.BaseDirectory, "Cart.Next.FSharp.Example01.Glue.dll")
        let glueAssembly = System.Reflection.Assembly.LoadFrom(gluePath)
        match glueAssembly.GetType("Dapr.Actors.Next.Generated.GeneratedActorRegistrationModule") with
        | null -> failwith "GeneratedActorRegistrationModule not found in Glue assembly"
        | moduleType ->
            match moduleType.GetMethod("Register", System.Reflection.BindingFlags.Static ||| System.Reflection.BindingFlags.NonPublic) with
            | null -> failwith "Register method not found"
            | registerMethod -> registerMethod.Invoke(null, null) |> ignore
        // Force the F# types to be loaded so the generated registry discovers them
        let _ = typeof<CartActor>
        ActorTestRuntime(fun services ->
            services.AddSingleton<IPricingClient, FakePricingClient>() |> ignore
            services.AddDaprActors(Action<DaprActorsOptions>(fun _ -> ())) |> ignore)

    [<Fact>]
    member this.Adding_an_item_updates_the_summary() = task {
        use runtime = CartActorTests.CreateRuntime()
        let cart = runtime.CreateActor<ICartActor>(ActorId.Create("cart-1"), "Cart")

        let add = cart.AddItem({ Sku = "sku-1"; Quantity = 2 }, CancellationToken.None)
        do! runtime.RunToIdle()
        do! add

        let read = cart.GetSummary(CancellationToken.None)
        do! runtime.RunToIdle()
        let! summary = read

        Assert.Equal(2, summary.ItemCount)
        Assert.Equal(25.00m, summary.Total)
        Assert.False(summary.Abandoned)
    }

    [<Fact>]
    member this.Advancing_virtual_time_fires_the_abandon_timer() = task {
        use runtime = CartActorTests.CreateRuntime()
        let cart = runtime.CreateActor<ICartActor>(ActorId.Create("idle-cart"), "Cart")

        let add = cart.AddItem({ Sku = "sku-1"; Quantity = 1 }, CancellationToken.None)
        do! runtime.RunToIdle()
        do! add

        runtime.Time.Advance(TimeSpan.FromMinutes(20.0))
        do! runtime.RunToIdle()

        let summaryTask = cart.GetSummary(CancellationToken.None)
        do! runtime.RunToIdle()
        let! summary = summaryTask
        Assert.True(summary.Abandoned)
    }

and FakePricingClient() =
    interface IPricingClient with
        member _.GetPriceAsync(sku: string, _: CancellationToken) : ValueTask<decimal> =
            ValueTask.FromResult(if sku = "sku-1" then 12.50m else 1.00m)
